torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
         else:
             return False
 
+    @contextlib.contextmanager
+    def set_state_dict_type(self, state_dict_type: StateDictType) -> Generator:
+        """
+        A context manager to set the state_dict_type of this FSDP module and
+        its descendant FSDP modules.
+        .. note:: This API should be called for only the root FSDP module.
+        .. note:: All the FSDP modules should have the same state_dict_type.
+        .. note:: The default state_dict_type is StateDictTyp.STATE_DICT.
+
+        Args:
+            state_dict_type (StateDictType): the desired state_dict_type to set.
+        """
+        prev_state_dict_type = self.state_dict_type
+        for module in self.modules():
+            if isinstance(module, FullyShardedDataParallel):
+                if module.state_dict_type != prev_state_dict_type:
+                    raise RuntimeError(
+                        "All FSDP module should the same state_dict_type."
+                    )
+                module.state_dict_type = state_dict_type
+        try:
+            yield
+        finally:
+            for module in self.modules():
+                if isinstance(module, FullyShardedDataParallel):
+                    module.state_dict_type = prev_state_dict_type
+
+    @staticmethod
+    def _post_state_dict_hook(
+        module: nn.Module,
+        state_dict: "OrderedDict[str, torch.Tensor]",
+        prefix: str,
+        *args: Any,
+    ) -> "OrderedDict[str, torch.Tensor]":
+        """
+        _post_state_dict_hook() is called after the state_dict() of this
+        FSDP module is executed. ``self.state_dict_type`` is used to decide
+        what postprocessing will be done.
+        """
+        self = cast(FullyShardedDataParallel, module)
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            flat_param = getattr(self.module, "flat_param", None)
+            full_numel = flat_param.full_numel
```

Comment:
do we need to have `flat_param is not None` assert before this line instead of on 730?